### PR TITLE
Start of ManipulationHandler rework

### PR DESCRIPTION
**Pull request by MenelvagorMilsom**
## Overview
- Adds `PrimitiveManipulator` component to experimental, which is ultimately intended to replace `ManipulationHandler`.
  - Changes some of the enums to use flags. This allowed me to simplify things within `ManipulationHandler`, but also exposes some more functionality to the user.
  - Removed the `ManipulationHandler` internal state machine, as it was too simple to be necessary and simply obfuscated pointer events.
  - Removed `RotateAboutObjectCenter` option, as it only behaved differently for one handed near interaction, and this different behaviour felt unnatural. 
  - Some minor cleanup of `ManipulationHandler` and the move/rotate/scale logic classes.
  - `PrimitiveManipulator` does not differ hugely from `ManipulationHandler` from the perspective of the user yet, but separating these classes now allows us to make bigger changes that will affect the user later.
- Improved the `ManipulationHandler` inspector and added the `PrimitiveManipulator` inspector.
  - Different sections are now contained in foldout sections.
  - Some sections are only displayed when certain conditions are met. These conditions are made clear in help boxes.
  - Added a warning for `ManipulationHandler` inspector, informing the user that `ManipulationHandler` will soon be removed. Also provided a button that will automatically upgrade the current object to use `PrimitiveInspector`.
- Added a new test for `ManipulationHandler` and new tests for `PrimitiveInspector`.

## Changes
- Fixes: #5464, #4396, #5776.

#### New Inspector
image

#### Upgrade Button
image

### CODE REVIEW

File: Assets/MixedRealityToolkit/Utilities/Physics/ManipulationMoveLogic.cs [hunk 1]
```diff
@@ -8,15 +8,15 @@
...
+    /// Implements a move logic that will move an object based on the initial position of 
+    /// the grab point relative to the pointer and relative to the object, and subsequent
+    /// changes to the pointer and the object's rotation
     /// 
     /// Usage:
     /// When a manipulation starts, call Setup.
     /// Call Update any time to update the move logic and get a new rotation for the object.
     /// </summary>
-    public class TwoHandMoveLogic
+    public class ManipulationMoveLogic
```
Comment (MenelvagorMilsom) [inline Assets/MixedRealityToolkit/Utilities/Physics/ManipulationMoveLogic.cs]: @wiwei should we move this, `TwoHandRotateLogic` and `TwoHandScaleLogic` to some internal namespace/location? They're not intended for users for normal usage and I've broken the public interface to these classes twice now. It would be nice if we could make it clear that these classes are subject to change.

Comment (thalbern) [reply Assets/MixedRealityToolkit/Utilities/Physics/ManipulationMoveLogic.cs]: can't we just define those classes internal instead of public?

Comment (thalbern) [reply Assets/MixedRealityToolkit/Utilities/Physics/ManipulationMoveLogic.cs]: there's no reason imho to have those public

File: Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs [hunk 2]
```diff
@@ -0,0 +1,744 @@
...
+            Scale = 1 << 2
+        };
+        public enum RotateInOneHandType
+        {
+            MaintainRotationToUser,
+            GravityAlignedMaintainRotationToUser,
+            FaceUser,
+            FaceAwayFromUser,
+            MaintainOriginalRotation,
+            RotateAboutGrabPoint
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs]: Looks like you removed RotateAboutObjectCenter, but actually I found that very useful for rotating things with one hand at a distance. I noticed that the behavior is broken in the 2.0 release, and filed #5957. Maybe you can fix #5957 in the new manipulationhandler, and we can discuss whether #5957 needs to be fixed for the old one (I posted a workaround that people can do in the mean time).

Comment (julenka) [reply Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs]: Note that doing #5957 core idea is for far manipulation you need to use the grip pose instead of the pointer pose -- the grip pose rotates with the palm, but the pointer pose uses the hand ray. For motion controllers it's similar -- grip pose is the rotation of the 6dof where pointer rotates based on where the 6dof is 'pointing' (not the same as grip pose).

File: Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs [hunk 3]
```diff
@@ -0,0 +1,744 @@
...
+        public float SmoothingAmount
+        {
+            get => smoothingAmount;
+            set => smoothingAmount = value;
+        }
+
+        #endregion Serialized Fields
+
+        #region Event handlers
+        public ManipulationEvent OnManipulationStarted = new ManipulationEvent();
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs]: Coding guidelines say to have private fields and public properties.

File: Assets/MixedRealityToolkit/Utilities/Physics/ManipulationMoveLogic.cs (hunk 1 above)
Comment (wiwei) [reply Assets/MixedRealityToolkit/Utilities/Physics/ManipulationMoveLogic.cs]: +1 on having these just be internal as well.

This is a breaking change in that we're removing the visibility of this, but this is a case where it should be a fairly obvious implementation detail of the the manipulation handler.

---
In reply to: [324756550](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5955#discussion_r324756550) [](ancestors = 324756550)

**Comment by wiwei** [comment]
>     public class TwoHandRotateLogic

I think similar to the comment that you ad in ManipulationMoveLogic, we should probably be consistent here with making sure that this (and TwoHandScaleLogic) are internal classes to manipulation handler (unless we want to make the statement that these actually should be pulled out to be more general classes i.e. moving them to some lower layer)

---
Refers to: Assets/MixedRealityToolkit/Utilities/Physics/TwoHandRotateLogic.cs:20 in 9b01488. [](commit_id = 9b01488c3968d810be1fdad03d7929d6518a74ec, deletion_comment = False)

File: Assets/MixedRealityToolkit/Utilities/Physics/TwoHandRotateLogic.cs [hunk 4]
```diff
@@ -25,19 +25,19 @@ public class TwoHandRotateLogic
         /// <summary>
         /// Setup the rotation logic.
         /// </summary>
-        public void Setup(Dictionary<uint, Vector3> handsPressedMap, Transform t, RotationConstraintType rotationConstraint)
+        public void Setup(Vector3[] handsPressedArray, Transform t, RotationConstraintType rotationConstraint)
```
Comment (wiwei) [inline Assets/MixedRealityToolkit/Utilities/Physics/TwoHandRotateLogic.cs]: >Vector3[] handsPressedArray [](start = 26, length = 27)

Not that this was defined well previously - it would be good to have a <param> doccomment about what the shape of this array is (i.e. it should be of length 2, where the new location is in the 1-th index, and the old location is in the 0-th index.

I think same goes for any public function in this class that takes a Vector3 array (I think they're all the same thing)

File: Assets/MixedRealityToolkit/Utilities/Physics/TwoHandScaleLogic.cs [hunk 5]
```diff
@@ -25,34 +25,32 @@ public class TwoHandScaleLogic
         /// <summary>
         /// Initialize system with source info from controllers/hands
         /// </summary>
-        /// <param name="handsPressedMap">Dictionary that maps inputSources to states</param>
+        /// <param name="handsPressedArray">Dictionary that maps inputSources to states</param>
         /// <param name="manipulationRoot">Transform of gameObject to be manipulated</param>
```
Comment (wiwei) [inline Assets/MixedRealityToolkit/Utilities/Physics/TwoHandScaleLogic.cs]: Nit: Update the doccomments (as per previous comment from last comment I just left - so meta - we should also document what this Vector3 should be -i.e. size two vector that has the old location in 0th index, new location in the 1th index)

File: Assets/MixedRealityToolkit/Utilities/Physics/TwoHandScaleLogic.cs [hunk 6]
```diff
@@ -25,34 +25,32 @@ public class TwoHandScaleLogic
...
-        public virtual Vector3 UpdateMap(Dictionary<uint, Vector3> handsPressedMap)
+        public virtual Vector3 UpdateMap(Vector3[] handsPressedArray)
         {
-            var ratioMultiplier = GetMinDistanceBetweenHands(handsPressedMap) / startHandDistanceMeters;
+            var ratioMultiplier = GetMinDistanceBetweenHands(handsPressedArray) / startHandDistanceMeters;
             return startObjectScale * ratioMultiplier;
         }
 
-        private float GetMinDistanceBetweenHands(Dictionary<uint, Vector3> handsPressedMap)
+        private float GetMinDistanceBetweenHands(Vector3[] handsPressedArray)
```
Comment (wiwei) [inline Assets/MixedRealityToolkit/Utilities/Physics/TwoHandScaleLogic.cs]: >GetMinDistanceBetweenHands [](start = 22, length = 26)

Not that you should actually do this:

[https://en.wikipedia.org/wiki/Closest_pair_of_points_problem](https://en.wikipedia.org/wiki/Closest_pair_of_points_problem)

File: Assets/MixedRealityToolkit.Examples/Demos/UX/ManipulationHandler/Scripts/ChangeManipulation.cs [hunk 7]
```diff
@@ -31,14 +32,28 @@ public void TryStopManipulation()
...
             if (manipulatedObject != null && collisionTrigger != null && manipulatedObjCollider != null)
             {
-                if (collisionTrigger.bounds.Intersects(manipulatedObjCollider.bounds))
+                if (!collisionTrigger.bounds.Intersects(manipulatedObjCollider.bounds))
                 {
-                    manipulatedObject.ForceEndManipulation();
+                    return;
+                }
+
+                var mh = manipulatedObject.GetComponent<ManipulationHandler>();
```
Comment (wiwei) [inline Assets/MixedRealityToolkit.Examples/Demos/UX/ManipulationHandler/Scripts/ChangeManipulation.cs]: >mh [](start = 20, length = 2)

Nit: May as well spell out the entire thing (i.e. manipulationHandler and primitiveManipulator).

File: Assets/MixedRealityToolkit.Examples/Demos/UX/ManipulationHandler/Scripts/ChangeManipulation.cs [hunk 8]
```diff
@@ -31,14 +32,28 @@ public void TryStopManipulation()
...
+                var pm = manipulatedObject.GetComponent<PrimitiveManipulator>();
+                if (mh != null)
+                {
+                    mh.ForceEndManipulation();
 
                     // move the object slightly away from the collision point so we can manipulate it again after this
                     Vector3 direction = collisionTrigger.bounds.center - manipulatedObjCollider.bounds.center;
                     manipulatedObject.transform.Translate(direction.normalized * 0.01f);
```
Comment (wiwei) [inline Assets/MixedRealityToolkit.Examples/Demos/UX/ManipulationHandler/Scripts/ChangeManipulation.cs]: If you think that there's any chance that we might add/change these three lines of code, consider changing the structure here to be something more like:

```csharp
if (mh != null || pm != null)
{
  if (mh != null)
  {
     mh.ForceEndManipulation()
  }
  else
  {
     pm.ForceEndManipulation()
  }

  // move the object slightly away from the collision point so we can manipulate it again after this
  Vector3 direction = collisionTrigger.bounds.center - manipulatedObjCollider.bounds.center;
  manipulatedObject.transform.Translate(direction.normalized * 0.01f);
}
```

If not then just resolve this conversation

Comment (wiwei) [reply Assets/MixedRealityToolkit.Examples/Demos/UX/ManipulationHandler/Scripts/ChangeManipulation.cs]: The reason being is that this type of copy paste has the tendency to introduce bugs where someone who is less experienced in the area will fix the "old" version of it (or "new" version) but not update the other, so you find out later and then the next person to fix it will be like "which one is actually correct?"

---
In reply to: [324907573](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5955#discussion_r324907573) [](ancestors = 324907573)

File: Assets/MixedRealityToolkit.SDK/Features/Input/Handlers/CursorContextManipulationHandler.cs [hunk 9]
```diff
@@ -36,8 +36,9 @@ private void Start()
         private void ManipulatorHoverEntered(ManipulationEventData manipEvent)
         {
             hoverCount++;
+            var mh = manipEvent.ManipulationSource.GetComponent<ManipulationHandler>();
```
Comment (wiwei) [inline Assets/MixedRealityToolkit.SDK/Features/Input/Handlers/CursorContextManipulationHandler.cs]: > var mh = manipEvent.ManipulationSource.GetComponent<ManipulationHandler>(); [](start = 11, length = 76)

Not really related to your current change, but do you know if this mh is always gonna be the same as the mh defined in Start()?

i.e. should we just cache mh in Start and use that to then do mh.ManipulationType?

I assume that what this code here is trying to do is ask questions about the manipulation handler on the same game object, not about the manipulation handler on some unrelated game object.

File: Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs [hunk 10]
```diff
@@ -0,0 +1,744 @@
...
+
+namespace Microsoft.MixedReality.Toolkit.Experimental.UI
+{
+    /// <summary>
+    /// This script allows for an object to be movable, scalable, and rotatable with one or two hands. 
+    /// You may also configure the script on only enable certain manipulations. The script works with 
+    /// both HoloLens' gesture input and immersive headset's motion controller input.
+    /// </summary>
+    [HelpURL("https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/README_ManipulationHandler.html")]
+    public class PrimitiveManipulator : MonoBehaviour, IMixedRealityPointerHandler, IMixedRealityFocusChangedHandler
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs]: How about ObjectManipulator instead of PrimitiveManipulator? PrimitiveManipulator sounds like it only works with primitives like sphere, cube.

File: Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs [hunk 11]
```diff
@@ -0,0 +1,744 @@
...
+
+        public MovementConstraintType ConstraintOnMovement
+        {
+            get => constraintOnMovement;
+            set => constraintOnMovement = value;
+        }
+        
+        [SerializeField]
+        [Tooltip("Check to enable frame-rate independent smoothing. ")]
+        private bool smoothingActive = true;
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs]: Since we are refactoring, could we also make smoothing params consistent with Solver? On Solvers like Overlap and RadialView, there'sa  "smoothing" checkbox, and then a "move lerp time" and "rotate lerp time" field. See Solver.cs:

```
public static Vector3 SmoothTo(Vector3 source, Vector3 goal, float deltaTime, float lerpTime)
        {
            return Vector3.Lerp(source, goal, lerpTime.Equals(0.0f) ? 1f : deltaTime / lerpTime);
        }
```

File: Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs [hunk 12]
```diff
@@ -0,0 +1,744 @@
...
+        }
+
+        private PointerData GetFirstPointer()
+        {
+            // We may be able to do this without allocating memory.
+            // Moving to a method for later investigation.
+            return pointerIdToPointerMap.Values.First();
+        }
+
+        private Quaternion ApplyConstraints(Quaternion newRotation)
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs]: This method actually doesn't seem so different from ManipulationHandler. Are you sure you can't just change the inspector? Seems like the only difference is that the HandMovementType and TwoHandManipulation type are flags. What else is different that requires us to have a new class?

File: Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs [hunk 13]
```diff
@@ -0,0 +1,744 @@
...
+            {
+                return;
+            }
+
+            // If we only allow one handed manipulations, check there is no hand interacting yet. 
+            if (manipulationType != HandMovementType.OneHanded || pointerIdToPointerMap.Count == 0)
+            {
+                uint id = eventData.Pointer.PointerId;
+                // Ignore poke pointer events
+                if (!eventData.used && !pointerIdToPointerMap.ContainsKey(id))
```
Comment (wiwei) [inline Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs]: >eventData.used [](start = 21, length = 14)

I think this is also an existing issue, but consider moving this !eventData.used check to the if statement at the beginning of this function (i.e. we don't do anything interesting if the eventData is used - I guess we MIGHT re-mark it as used, but since it's already used, that's also a no-op).

File: Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs [hunk 14]
```diff
@@ -0,0 +1,744 @@
...
+                {
+                    // cache start ptr grab point
+                    pointerIdToPointerMap.Add(id, new PointerData(eventData.Pointer, eventData.Pointer.Result.Details.Point));
+
+                    // Call manipulation started handlers
+                    var handsPressedCount = pointerIdToPointerMap.Count;
+                    if (manipulationType.HasFlag(HandMovementType.TwoHanded) && handsPressedCount > 1)
+                    {
+                        if (!manipulationType.HasFlag(HandMovementType.OneHanded))
```
Comment (wiwei) [inline Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs]: >!manipulationType.HasFlag(HandMovementType.OneHanded)) [](start = 28, length = 54)

I think that this line makes some assumptions about ordering with the else if statement down below (i.e. that the one handed manipulation thing would have been engaged first - totally valid, though this is also somewhat of a state machine in itself - thoughts on just having:

```csharp
if (!isManipulationStarted)
{
  HandleManipulationStarted();
}
```

File: Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs [hunk 15]
```diff
@@ -0,0 +1,744 @@
...
+                // This is due to us reacting to both "Select" and "Grip" events.
+                eventData.Use();
+            }
+        }
+
+        public void OnPointerDragged(MixedRealityPointerEventData eventData)
+        {                    
+            // Call manipulation updated handlers
+            var handsPressedCount = pointerIdToPointerMap.Count;
+            if (manipulationType.HasFlag(HandMovementType.OneHanded) && handsPressedCount == 1)
```
Comment (wiwei) [inline Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs]: >manipulationType.HasFlag(HandMovementType.OneHanded) && handsPressedCount == 1 [](start = 16, length = 78)

There's a decent amount of copy/paste instances of this + the two hand case (i.e. twohanded + > 1) - consider having a helper function like IsOneHandedManipulationActive() or IsTwoHandedManipulationActive() to avoid the copy paste.

Comment (wiwei) [reply Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs]: Active actually may be a bad word because it's also used in the pervious function to also start the manipulation as well, but yeahhhh

---
In reply to: [324915287](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5955#discussion_r324915287) [](ancestors = 324915287)

Comment (wiwei) [reply Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs]: Or maybe not, maybe this isn't used too much down below (still reading)

---
In reply to: [324915398](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5955#discussion_r324915398) [](ancestors = 324915398,324915287)

**Comment by SimonDarksideJ** [comment]
For this change, can you elaborate on the deficiencies for the existing ManipulationHandler and describe in what way this new handler improves the situation and ultimate goals.

Lists of new features are good, but it's better to understand why this is needed

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/PrimitiveManipulatorTests.cs [hunk 16]
```diff
@@ -0,0 +1,1000 @@
...
+            Assert.AreEqual(3, manipulationStartedCount);
+            Assert.AreEqual(2, manipulationEndedCount);
+
+            yield return rightHand.SetGesture(ArticulatedHandPose.GestureId.Open);
+            Assert.AreEqual(3, manipulationStartedCount);
+            Assert.AreEqual(2, manipulationEndedCount);
+
+            yield return leftHand.SetGesture(ArticulatedHandPose.GestureId.Open);
+            Assert.AreEqual(3, manipulationStartedCount);
+            Assert.AreEqual(3, manipulationEndedCount);
```
Comment (wiwei) [inline Assets/MixedRealityToolkit.Tests/PlayModeTests/PrimitiveManipulatorTests.cs]: Now that the test has been already written, this is more of a consideration for next time, but in general having tests who state assertions depend on previous state (i.e. the fact that things start at 2 in this test case are because the previous "cases" already exist). This makes it hard to do things like delete things from earlier parts in the test (imagine we were to add a new test case in the middle after the first case and before the second case, we'd have to change up everything after it).

Where in reality, these really are separate test cases (i.e. PrimitiveManipulatorOneHandedEvents, PrimitiveManipulatorTwoHandedEvents, PrimitiveManipulatorOneAndTwoHandedEvents)

This is something that may not actually matter but when it does, it's something that you would wish that you had had structured differently.

File: Assets/MixedRealityToolkit.SDK/Features/Input/Events/ManipulationEventData.cs [hunk 17]
```diff
@@ -13,7 +13,7 @@ public class ManipulationEventData
         /// <summary>
         /// Source of ManipulationEvent.
         /// </summary>
-        public ManipulationHandler ManipulationSource { get; set; }
+        public GameObject ManipulationSource { get; set; }
```
Comment (wiwei) [inline Assets/MixedRealityToolkit.SDK/Features/Input/Events/ManipulationEventData.cs]: >GameObject [](start = 15, length = 10)

Thoughts on keeping the old thing around with the Obsolete attribute?

File: Assets/MixedRealityToolkit.Examples/Demos/UX/ManipulationHandler/Scenes/PrimitiveManipulatorExample.unity.meta [hunk 18]
```diff
@@ -0,0 +1,7 @@
+fileFormatVersion: 2
+guid: 078c2a355a7d1b74995eb21f9ba3dc61
```
[the remark this answers is not in the document]
Comment (MenelvagorMilsom) [reply Assets/MixedRealityToolkit.Examples/Demos/UX/ManipulationHandler/Scenes/PrimitiveManipulatorExample.unity.meta]: Oops, myself and @thalbern were unsure if we should add an experimental directory to the Examples package, but it looks like it already exists. I'll move this.

File: Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs (hunk 2 above)
Comment (MenelvagorMilsom) [reply Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs]: Ah ok, I removed because I had only observed the broken behaviour. I'll fix and add this option back.

File: Assets/MixedRealityToolkit/Utilities/Physics/TwoHandScaleLogic.cs (hunk 6 above)
Comment (MenelvagorMilsom) [reply Assets/MixedRealityToolkit/Utilities/Physics/TwoHandScaleLogic.cs]: I did think this, but it may be a bit excessive when generally we're talking about no more than two points, haha.

File: Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs (hunk 12 above)
Comment (MenelvagorMilsom) [reply Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs]: Yeah, there isn't a huge amount of difference yet. I separated out a lot of WIP changes so that I'm not pushing everything into one monolithic PR. I figured that it's better to separate these classes now so that we can make the changes we wanted to make in the next couple of iterations without having to worry about breaking changes.

I was going to look at constraints next, and making them consistent with BoundingBox. I already have most of the changes locally.

**Comment by MenelvagorMilsom** [comment]
> For this change, can you elaborate on the deficiencies for the existing ManipulationHandler and describe in what way this new handler improves the situation and ultimate goals.
> 
> Lists of new features are good, but it's better to understand why this is needed

@SimonDarksideJ So the main thing addressed in this PR is that the ManipulationHandler inspector properties. At the moment they imply behaviour is possible, without explaining what you need to change about the object in order to actually enable that behaviour. 

Release behaviour, for example, implies that the object can maintain it's velocity on release, without explaining to the user that you need a Rigidbody component on the object.

The reason why I feel like we need to separate these changes into their own class now, is that there are other planned changes to ManipulationHandler, some of which I already have a good portion of the work done for, that will be severe breaking changes. Separating now and adding to experimental means that we can continue refactoring this class without worrying as much about breaking changes. 

For example, I have a lot of changes locally refactoring contraints, in the hopes of making them more consistent between ManipulationHandler and BoundingBox. I decided to pause work on those changes so I could get this PR out, rather than having one Monolithic PR.

On top of all this there was some cleanup I felt was necessary to make ManipulationHandler more readable and approachable for developers, these will aid the future changes I mentioned.

File: Assets/MixedRealityToolkit.SDK/Inspectors/UX/ManipulationHandler/ManipulationHandlerInspector.cs [hunk 19]
```diff
@@ -4,23 +4,261 @@
...
+
+            // reset foldouts style
+            style.fontStyle = previousStyle;
+
+            serializedObject.ApplyModifiedProperties();
+        }
+
+        private void Migrate()
+        {
+            var mh1 = target as ManipulationHandler;
```
Comment (thalbern) [inline Assets/MixedRealityToolkit.SDK/Inspectors/UX/ManipulationHandler/ManipulationHandlerInspector.cs]: I think it would be better to move the Migrate function into a MigrationUtility or other public space, so we can use it for a batch conversion as well. 
Not part of this PR: might be good to add a Migration Utility Window for MRTK where you can trigger upgrading to the new UX elements in current scene / all scenes / prefabs etc.

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/ObjectManipulatorTests.cs [hunk 20]
```diff
@@ -0,0 +1,1231 @@
...
+            diffAngle = Quaternion.Angle(testObject.transform.rotation, Quaternion.Euler(0, 0, testRotation) * initialQuaternion);
+            Assert.IsTrue(Mathf.Approximately(diffAngle, 0.0f), "constraint on Z axis did not lock axis correctly (world space)");
+
+            yield return hand.SetRotation(Quaternion.identity, numRotSteps);
+            diffAngle = Quaternion.Angle(testObject.transform.rotation, initialQuaternion);
+            Assert.IsTrue(Mathf.Approximately(diffAngle, 0.0f), "constraint on Z axis did not lock axis correctly (world space)");
+            
+            ///*********************************/
+            ///*** TEST LOCAL SPACE ROTATION ***/
+            ///*********************************/
```
[the remark this answers is not in the document]
Comment (MenelvagorMilsom) [reply Assets/MixedRealityToolkit.Tests/PlayModeTests/ObjectManipulatorTests.cs]: Yes, already have it running and passing alongside the constraint changes.
